feat(editor): `:bcd` changes buffer-local directory

msaher opened the pull request:
Closes #33318 

## Problem
No way to set a buffer-local directory.
Use-cases:
- "Root dir" for LSP (and the ["project" concept](https://github.com/neovim/neovim/issues/8610)).
- `:terminal` OSC 7

## Solution
Add `:bcd` command.

- buffer-local directories should have less priority than window-local ones, and more priority than tab-local ones.
- extend `getcwd()` to take a third argument such that `getcwd(-1, -1, bunfr)` returns the buffer-local working directory.

msaher committed:
feat(editor): `:bcd` changes buffer-local directory

Problem:
No way to set a buffer-local directory.
Use-cases:
- "Root dir" for LSP (and the "project" concept).
- `:terminal` OSC 7

Solution:
Add `:bcd` command.

- Extend `getcwd()` to take a third arg; `getcwd(-1, -1, bunfr)` returns
  the buffer-local working directory.
- Buffer-local directories have less priority than window-local
  ones, and more priority than tab-local ones.

Co-authored-by: Justin M. Keyes <justinkz@gmail.com>

justinmk commented:
Looks like there was some merge/rebase confusion here, since this has unrelated changes from already-merged PRs.

Given the discussion in https://github.com/neovim/neovim/issues/8610, it does seem like this could be a necessary feature. Need feedback from others.

Note also that a requirement is to have an ergonomic way to get the buffer's "current directory". At least via `getcwd()`, but also maybe a buffer-local option or variable? (Need to hear comments on that.)

msaher commented:
> Looks like there was some merge/rebase confusion here, since this has unrelated changes from already-merged PRs.

Apologies. I'm kinda new to this. Do I rebase on top of master and force push? I think I was using an outdated master branch.

> ergonomic way to get the buffer's "current directory"

This PR extends `getcwd()` to take a third argument such that `getcwd(-1, -1, bunfr)` returns the buffers' working directory.

 > maybe buffer-local option or variable?

`:lcd` does not set window-local options, so it will be inconsistent if `:bcd` sets buffer-local options.

justinmk commented:
> Do I rebase on top of master and force push?

yes

msaher commented:
I think I should give an overview of the changes.

This PR adds `:bcd` command to change buffer-local directories. This is done by adding a new entry to the `CdScope` enum. The new hierarchy is `window-local -> buffer-local -> tab-local -> global`. The `getcwd()` and `haslocaldir()` functions will follow this hierarchy.

It's my personal opinion that buffer-local directories should have less priority than window-local ones, and more priority than tab-local ones. I'm open to discuss this if there are objections.  

The `:cd`, `:tcd` and `:lcd` family of commands do not set any options, so following that tradition, `:bcd` does not set a buffer-local option. To query the current directory, use `getcwd()`.

This PR updates `getcwd()` and `haslocaldir()` to support buffer-local directories. This new signature is `getcwd({winnr}, {tabnr}, {bufnr})`. When `{bufnr}` is provided, `{winnr}` and `{tabnr}` MUST be -1. Otherwise, it's an error (because passing a window or tab doesn't make sense). When `{bufnr}` is missing, and `{winnr}` or `{tabnr}` is provided, the behavior is the same as the old versions. `haslocaldir()` has similar updates.

I'd like some feedback on this potentially confusing edge case this:

```vimscript
getcwd(-1)        "a) get tab-local directory. fine.
getcwd(-1, 0)     "b) get tab-local directory. fine
getcwd(-1, -1)    "c) get global directory, NOT buffer-local directory. Isn't confusing?
getcwd(-1, -1, 0) "d) get buffer-local directory. fine.
```
  
One could argue that `c)` should return a buffer-local directory rather than a global one because its consistent with `a)` and the pattern that -1 means give me the "next" scope. However, the current documentation of `:h getcwd` says

> If both `{winnr}` and `{tabnr}` are -1 the global working directory is returned.

So changing `c)` to return a global directory would technically be a breaking change.

Implementation wise, vim calls `win_fix_current_dir()` to update the directory when navigating between windows. The introduction of buffer-local directories means switching to a different buffer in the same window could trigger a directory change. Thus, `win_fix_current_dir()` has been changed into `fix_current_dir(bool caused_by_win)`. 

The `DirChangedPre` and `DirChanged` auto commands are now aware of buffer-local directories and their `scope` event key may now take the value of `"buffer"`.

Finally, `:bcd` is restored by `:mksession`. This is done by creating a autocmd that runs `++once` on `BufEnter` to call `:bcd`.

justinmk commented:
This looks rather complete. I don't have any objections, waiting to hear other comments...

Since you are thinking about this, is there an example usage you can provide in lsp.txt? And any other concrete usage examples will help.

To fill out the "project" concept (not this PR), we may need a function that makes it easy for plugins to "get all the buffers for a given workdir" (pseudoname [get_project_bufs](https://github.com/neovim/neovim/issues/8610#issuecomment-1367994885)). That also may involve "nested" projects...

> So changing `c)` to return a global directory would technically be a breaking change.

Yeah, changing that would be very expensive.



> buffer-local directories should have less priority than window-local ones, and more priority than tab-local ones.

Makes sense.

zeertzjq commented on src/nvim/eval.lua:
```diff
@@ -5172,14 +5180,15 @@ M.funcs = {
+      -1 and {bufnr} has set a local path via |:bcd|, otherwise 0.
 
-      Tabs and windows are identified by their respective numbers,
+      Tabs, windows and buffers are identified by their respective numbers,
```
Line length over limit

msaher replied:
Good catch. Am I supposed to amend and force push or create a new commit?

msaher commented on runtime/doc/editing.txt:
```diff
@@ -1367,10 +1368,16 @@ exist, the next-higher scope in the hierarchy applies.
 							*:tch* *:tchdir*
 :tch[dir][!]		Same as |:tcd|.
 
+							*:bc* *:bcd*
+:bc[d][!] {path}	Like |:bcd|, but only set the current directory for the
+			current buffer.  The current directory for other
+			windows or buffers or tabs is not changed.
+
+
```
`:h chdir()` must be updated as well. Forgot about that

msaher commented on test/functional/ex_cmds/cd_spec.lua:
```diff
@@ -271,14 +392,19 @@ for _, cmd in ipairs { 'getcwd', 'haslocaldir' } do
       eq(err474, exc_exec('call ' .. cmd .. '(-2)'))
     end)
     local err5001 = 'Vim(call):E5001: Higher scope cannot be -1 if lower scope is >= 0.'
+    local err5006 = 'Vim(call):E5006: Window and tab scope must be -1 when using buffer scope'
```
forgot to test `err5007`

zeertzjq commented on runtime/doc/editing.txt:
```diff
@@ -1367,10 +1368,16 @@ exist, the next-higher scope in the hierarchy applies.
+:bc[d][!] {path}	Like |:bcd|, but only set the current directory for the
+			current buffer.  The current directory for other
+			windows or buffers or tabs is not changed.
+
```
```suggestion
```

msaher commented on src/nvim/ex_cmds.c:
```diff
@@ -2709,6 +2709,12 @@ int do_ecmd(int fnum, char *ffname, char *sfname, exarg_T *eap, linenr_T newlnum
     redraw_curbuf_later(UPD_NOT_VALID);  // redraw this buffer later
   }
 
+  // If editing a buffer in the current window, make sure to update to the
+  // buffer's working directory.
+  if (oldwin == curwin) {
+    fix_current_dir(false);
+  }
+
```
This will not work with `:enew` because it will call `buflist_new()` above with the `BLN_CURBUF` flag, and that will call `buf_freeall()`, freeing `curbuf->b_localdir` and setting it to `NULL`. Effectively ignoring the buffer-local directory.

justinmk replied:
can avoid the problem entirely by dropping the "sticky"/inherit behavior: https://github.com/neovim/neovim/pull/33320/changes/514a3ded9fdbe2725ecc9a367d6591ac70a5d2f9

msaher commented:
Another good example to add in the docs is to replace the snippet in `:h terminal-osc7` to use `:bcd` instead of `:cd`.

justinmk commented:
A use-case for buffer-local CWD: ~~it removes the need to store CWD in the `term://` URI.~~

@msaher do you plan to continue this? needs a rebase.

Willaaaaaaa commented on runtime/doc/editing.txt:
```diff
@@ -1393,10 +1394,15 @@ exist, the next-higher scope in the hierarchy applies.
 :tch[dir][!]		Same as |:tcd|.
 
+							*:bc* *:bcd*
+:bc[d][!] {path}	Like |:bcd|, but only set the current directory for the
```
```suggestion
:bc[d][!] {path}	Like |:cd|, but only set the current directory for the
```

Willaaaaaaa commented on runtime/doc/vimfn.txt:
```diff
@@ -1019,13 +1019,16 @@ chdir({dir} [, {scope}])                                               *chdir()*
 		If {scope} is present, changes the current working directory
 		for the specified scope:
 		    "window"	Changes the window local directory.  |:lcd|
+		    "buffer"	Changes the window local directory.  |:bcd|
```
```suggestion
		    "buffer"	Changes the buffer local directory.  |:bcd|
```

Willaaaaaaa commented on runtime/doc/editing.txt:
```diff
@@ -1393,10 +1394,15 @@ exist, the next-higher scope in the hierarchy applies.
+:bc[d][!] {path}	Like |:bcd|, but only set the current directory for the
+			current buffer.  The current directory for other
+			windows or buffers or tabs is not changed.
+
```
Maybe we should also add docs for `:bcd-` and `:bch` for consistency : )

justinmk commented:
rebased.

justinmk commented on test/functional/ex_cmds/cd_spec.lua:
```diff
@@ -239,6 +297,137 @@ for _, cmd in ipairs { 'cd', 'chdir' } do
+      command('bd') -- delete buffer
+    end)
+
+    it('makes :new use the buffer-local directory', function()
```
This seems contrary to one of the key benefits of buffer-local CWD, which is that it *shouldn't* "inherit" buf-local CWD (unlike win-local / `:lcd` CWD). 

@msaher was this intentional, what's the use-case?

msaher replied:
Not fully intentional. I assumed a buffer-local CWD must be inherited (because window-local and tab-local CWDs are inherited), but now that you mention it I can see why it shouldn't be inherited.

justinmk replied:
ok thanks, let's try it without "inherit" for now. then can revert later if we find a strong reason to "inherit".

I will push a commit in a bit... https://github.com/neovim/neovim/pull/33320/changes/514a3ded9fdbe2725ecc9a367d6591ac70a5d2f9

justinmk committed:
refactor(editor): cleanup change-directory (:bcd) logic

justinmk committed:
fix(cwd): validate getcwd(…, -1)

msaher commented on src/nvim/ex_cmds.c:
```diff
@@ -2937,6 +2937,24 @@ int do_ecmd(int fnum, char *ffname, char *sfname, exarg_T *eap, linenr_T newlnum
+    if (old_curbuf.br_buf->b_prevdir != NULL) {
+      curbuf->b_prevdir = xstrdup(old_curbuf.br_buf->b_prevdir);
+    }
+  }
```
@justinmk, this is the inheritance logic.

justinmk committed:
fix(:bcd): do not "inherit" buffer-local dir

Problem:
Buffer-local CWD (:bcd) is "sticky", similar to window-local CWD (:lcd).
But this contradicts one of its main benefits: per-buffer "project root"
for LSP, OSC7.

Other problems:
- A buffer created with :edit/:enew/:new silently inherits b_localdir
  (and b_prevdir) from the previous buffer.
- curbuf_reusable() refuses to recycle a scratch buffer that has
  `b_localdir`.
- After :new/:vnew/:tabnew the CWD sticks to previous buffer's
  `b_localdir` even though the new curbuf has none, so :new is not
  equivalent to ":split | enew", and getcwd() disagrees with
  haslocaldir().
- Requires "which buffer spawned this buffer" semantics that no other
  buffer-local state has.

Solution:
Drop sticky/inherit behavior of buffer-local CWD (:bcd).

- do_ecmd: always apply the new curbuf's dir (`fix_current_dir`), like
  `do_autochdir` already does. :tabnew from a :bcd buffer now reverts to
  global CWD (and fires DirChanged), same as :tabnew from a :lcd window.
- curbuf_reusable(): recycling a scratch buffer frees its b_localdir.

To get sticky/inherit behavior of CWD, use `:lcd`.

justinmk committed:
fix(types): emmylua warning about `params` field
